Fix missing stm32u3 timers

Richy42 opened the pull request:
stm32u3: Add timers support (pwm, counter)

erwango commented on tests/drivers/pwm/pwm_api/boards/nucleo_u385rg_q.overlay:
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2021 Denx Software Engineering GmbH
```
Is this your copyright ?
Otherwise, please use yours. Apply everywhere.

erwango commented on tests/drivers/pwm/pwm_api/boards/nucleo_u385rg_q.overlay:
```diff
@@ -24,17 +24,3 @@
 	};
 };
-
-&timers5 {
```
Please fix initial commit directly

erwango reviewed (changes requested):
Thanks for this. Few comments but otherwise should be good.

I'd suggest to rename PR title as follows:
stm32u3: Add timers support (pwn, counter)

Richy42 commented on dts/arm/st/u3/stm32u3.dtsi:
```diff
@@ -433,6 +434,179 @@
+			compatible = "st,stm32-timers";
+			reg = <0x40012c00 0x400>;
+			clocks = <&rcc STM32_CLOCK(APB2, 11)>,
+				<&rcc STM32_SRC_PCLK1 NO_SEL>;
```
This was my initial guess, but in pwm_stm32.c function pwm_stm32_init a change after release 4.2 was introduced:

```
	if (cfg->pclk_len > 1) {
		/* Enable Timer clock source */
		r = clock_control_configure(clk, (clock_control_subsys_t)&cfg->pclken[1], NULL);
		if (r != 0) {
			LOG_ERR("Could not configure clock (%d)", r);
			return r;
		}

		r = clock_control_get_rate(clk, (clock_control_subsys_t)&cfg->pclken[1], &tim_clk);
		if (r < 0) {
			LOG_ERR("Timer clock rate get error (%d)", r);
			return r;
		}
	} else {
		LOG_ERR("Timer clock source is not specified");
		return -EINVAL;
	}
```

that change expects two clock definitions otherwise it throws an error.

One way for a solution for me would have been to remove the else path to get it working.
But then I thought that it was intentional and the correct way is to have a NO_SEL clock and add the check I did in the LL module as that was done for other MCUs already.

Initially I added the timer nodes based on the 4.2 release and there it worked fine with only one clock, when I switched to main the problems started ;-)

So, what would be the best way to go?

gautierg-st replied:
For U3, the timer source clock is PCLKx if APBx prescaler is 1, and 2*PCLKx otherwise.
So the kernel clock field of all timers should take `STM32_SRC_TIMPCLK1` or `STM32_SRC_TIMPCLK2`.
These values need to be defined in `zephyr\include\zephyr\dt-bindings\clock\stm32u3_clock.h` (like it's done for practically all other series).

erwango commented on tests/drivers/counter/counter_basic_api/boards/nucleo_u385rg_q.overlay:
```diff
@@ -0,0 +1,62 @@
+&timers1 {
```
Please add copyright and license info.
We need to fix the others, but let's make it correct for this one.

Richy42 replied:
If there is already a copyright line e.g. from STM, is the correct way to add another one?

Richy42 replied:
but keep existing ones or replace with the above?

etienne-lms reviewed (commented):
I missed to post a geneeric comment. For consistency of the commit messages header lines, could you use standard prefixes, e.g.:
```patch
-dts: timer definitions missing for stm32u3
+dts: arm: st: stm32u3: define timer nodes
```

```patch
-drivers: fix clock configuration for stm32u3
+drivers: clock_control: stm32: fix configuration for stm32u3
```
```patch
-boards: add pwm-leds and add support for nucleo_u385rg_q
+boards: st: nucleo_u385rg_q: add PWM LEDs
```
```patch
-tests: add pwm test to board nucleo_u385rg_q
+tests: drivers: pwm: pwm_api: add nucleo_u385rg_q
```
```patch
-tests: add counter test to board nucleo_u385rg_q
+tests: drivers: counter: counter_basic_api: add nucleo_u385rg_q
```
```patch
-samples: added board nucleo_u385rg_q to the pwm blinky sample
+samples: basic: blinky_pwm: add nucleo_u385rg_q
```

gautierg-st commented on dts/arm/st/u3/stm32u3.dtsi:
```diff
@@ -433,6 +434,179 @@
+			compatible = "st,stm32-timers";
+			reg = <0x40014400 0x400>;
+			clocks = <&rcc STM32_CLOCK(APB2, 17)>,
+				 <&rcc STM32_SRC_TIMPCLK1 NO_SEL>;
```
Ditto

gautierg-st commented on dts/arm/st/u3/stm32u3.dtsi:
```diff
@@ -433,6 +434,179 @@
+			compatible = "st,stm32-timers";
+			reg = <0x40014800 0x400>;
+			clocks = <&rcc STM32_CLOCK(APB2, 18)>,
+				 <&rcc STM32_SRC_TIMPCLK1 NO_SEL>;
```
Ditto

gautierg-st reviewed (changes requested):
Some clock nodes use the wrong kernel clock.

Nit: could you also move your clock commit first before the dtsi changes (so that chronologically the symbol in the dtsi are always defined)?

Richy42 committed:
drivers: clock_control: stm32: fix configuration for stm32u3

A change was introduced in how stm32 timers clocks are defined
to have multiple clock definitions.
This change was not handled in the stm32u3 series driver
(which had no timers defined in the dts at all).
This fix ports the required code to the stm32u3 mcu.

Signed-off-by: Richard Skriwanek <richy@fnc.at>

Richy42 committed:
dts: arm: st: stm32u3: define timer nodes

The dtsi file for STM32U3 was missing the timer definitions.
Fixed by adding all missing timers.

Signed-off-by: Richard Skriwanek <richy@fnc.at>

Richy42 committed:
boards: st: nucleo_u385rg_q: add PWM LEDs

The nucleo_u385rg_q board had no pwm-led defined.
Added the relevant led and timer definitions.
Additionally the pwm and counter are added
to supported functions.

Signed-off-by: Richard Skriwanek <richy@fnc.at>

Richy42 committed:
tests: drivers: pwm: pwm_api: add nucleo_u385rg_q

PWM API test were not available for the nucleo_u385rg_q board.
Added the nucleo_u385rg_q board overlay to the pwm_api test.

Signed-off-by: Richard Skriwanek <richy@fnc.at>

Richy42 committed:
tests: drivers: counter: counter_basic_api: add nucleo_u385rg_q

Counters API test were not available for the nucleo_u385rg_q board.
Added the nucleo_u385rg_q board overlay to the
counter_basic_api test.

Signed-off-by: Richard Skriwanek <richy@fnc.at>

Richy42 committed:
samples: basic: blinky_pwm: add nucleo_u385rg_q

The blinky_pwm test was not available for the nucleo_u385rg_q board.
Added the board overlay to the blinky_pwm sample.

Signed-off-by: Richard Skriwanek <richy@fnc.at>

gautierg-st reviewed (approved):
LGTM. Thanks

FRASTM commented:
more pertinent for next release

gautierg-st commented:
@nordic-krch Gentle ping for review. Or else reassign to @erwango or myself. Thanks
